File: hexatuneapp/ios/Runner/AudioDeviceDetector.swift
// SPDX-FileCopyrightText: 2025 hexaTune LLC
// SPDX-License-Identifier: MIT

import Flutter
import AVFoundation

/// Detects wired and wireless (including BLE) audio device connections
/// using `AVAudioSession.routeChangeNotification`.
///
/// Communicates with the Dart side via a `FlutterMethodChannel` for queries
/// and a `FlutterEventChannel` for real-time connection/disconnection events.
class AudioDeviceDetector: NSObject, FlutterStreamHandler {
    private let methodChannel: FlutterMethodChannel
    private let eventChannel: FlutterEventChannel
    private var eventSink: FlutterEventSink?

    init(messenger: FlutterBinaryMessenger) {
        methodChannel = FlutterMethodChannel(
            name: "com.hexatune/audio_device_detector",
            binaryMessenger: messenger
        )
        eventChannel = FlutterEventChannel(
            name: "com.hexatune/audio_device_events",
            binaryMessenger: messenger
        )
        super.init()
    }

    func start() {
        methodChannel.setMethodCallHandler { [weak self] call, result in
            self?.onMethodCall(call, result: result)
        }
        eventChannel.setStreamHandler(self)
    }

    // MARK: - FlutterStreamHandler

    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        eventSink = events
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(handleRouteChange),
            name: AVAudioSession.routeChangeNotification,
            object: AVAudioSession.sharedInstance()
        )
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        NotificationCenter.default.removeObserver(
            self,
            name: AVAudioSession.routeChangeNotification,
            object: AVAudioSession.sharedInstance()
        )
        eventSink = nil
        return nil
    }

    // MARK: - Route change handling

    @objc private func handleRouteChange(_ notification: Notification) {
        guard let userInfo = notification.userInfo,
              let reasonValue = userInfo[AVAudioSessionRouteChangeReasonKey] as? UInt,
              let reason = AVAudioSession.RouteChangeReason(rawValue: reasonValue) else {
            return
        }

        switch reason {
        case .newDeviceAvailable:
            let route = AVAudioSession.sharedInstance().currentRoute
            if hasWiredHeadset(in: route) {
                eventSink?("wired_connected")
            } else if hasWirelessHeadset(in: route) {
                eventSink?("wireless_connected")
            }

        case .oldDeviceUnavailable:
            if let previousRoute = userInfo[AVAudioSessionRouteChangePreviousRouteKey] as? AVAudioSessionRouteDescription {
                if hasWiredHeadset(in: previousRoute) {
                    eventSink?("wired_disconnected")
                } else if hasWirelessHeadset(in: previousRoute) {
                    eventSink?("wireless_disconnected")
                }
            }

        default:
            break
        }
    }

    // MARK: - Method channel

    private func onMethodCall(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "getCurrentState":
            let route = AVAudioSession.sharedInstance().currentRoute
            result([
                "wired": hasWiredHeadset(in: route),
                "wireless": hasWirelessHeadset(in: route),
            ])
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - Device type helpers

    private func hasWiredHeadset(in route: AVAudioSessionRouteDescription) -> Bool {
        return route.outputs.contains { $0.portType == .headphones || $0.portType == .usbAudio }
    }

    private func hasWirelessHeadset(in route: AVAudioSessionRouteDescription) -> Bool {
        return route.outputs.contains {
            $0.portType == .bluetoothA2DP ||
            $0.portType == .bluetoothHFP ||
            $0.portType == .bluetoothLE
        }
    }
}

File: hexatuneapp/android/app/src/main/kotlin/com/hexatune/hexatuneapp/AudioDeviceDetector.kt
// SPDX-FileCopyrightText: 2025 hexaTune LLC
// SPDX-License-Identifier: MIT

package com.hexatune.hexatuneapp

import android.content.Context
import android.media.AudioDeviceCallback
import android.media.AudioDeviceInfo
import android.media.AudioManager
import android.os.Build
import android.os.Handler
import android.os.Looper
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

/**
 * Detects wired and wireless (including BLE) audio device connections
 * using [AudioManager.registerAudioDeviceCallback].
 *
 * Communicates with the Dart side via a [MethodChannel] for queries
 * and an [EventChannel] for real-time connection/disconnection events.
 */
class AudioDeviceDetector(
    private val context: Context,
    private val methodChannel: MethodChannel,
    private val eventChannel: EventChannel,
) {
    private var eventSink: EventChannel.EventSink? = null
    private val audioManager = context.getSystemService(Context.AUDIO_SERVICE) as AudioManager
    private val mainHandler = Handler(Looper.getMainLooper())

    private val audioDeviceCallback = object : AudioDeviceCallback() {
        override fun onAudioDevicesAdded(addedDevices: Array<out AudioDeviceInfo>) {
            for (device in addedDevices) {
                if (!device.isSink) continue
                when {
                    device.isWiredHeadset() -> sendEvent("wired_connected")
                    device.isWirelessHeadset() -> sendEvent("wireless_connected")
                }
            }
        }

        override fun onAudioDevicesRemoved(removedDevices: Array<out AudioDeviceInfo>) {
            for (device in removedDevices) {
                if (!device.isSink) continue
                when {
                    device.isWiredHeadset() -> sendEvent("wired_disconnected")
                    device.isWirelessHeadset() -> sendEvent("wireless_disconnected")
                }
            }
        }
    }

    fun start() {
        methodChannel.setMethodCallHandler(::onMethodCall)

        eventChannel.setStreamHandler(object : EventChannel.StreamHandler {
            override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
                eventSink = events
                audioManager.registerAudioDeviceCallback(audioDeviceCallback, mainHandler)
            }

            override fun onCancel(arguments: Any?) {
                audioManager.unregisterAudioDeviceCallback(audioDeviceCallback)
                eventSink = null
            }
        })
    }

    fun stop() {
        audioManager.unregisterAudioDeviceCallback(audioDeviceCallback)
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        eventSink = null
    }

    private fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "getCurrentState" -> {
                val devices = audioManager.getDevices(AudioManager.GET_DEVICES_OUTPUTS)
                val wired = devices.any { it.isSink && it.isWiredHeadset() }
                val wireless = devices.any { it.isSink && it.isWirelessHeadset() }
                result.success(mapOf("wired" to wired, "wireless" to wireless))
            }
            else -> result.notImplemented()
        }
    }

    private fun sendEvent(event: String) {
        mainHandler.post { eventSink?.success(event) }
    }

    private fun AudioDeviceInfo.isWiredHeadset(): Boolean {
        return type == AudioDeviceInfo.TYPE_WIRED_HEADPHONES ||
                type == AudioDeviceInfo.TYPE_WIRED_HEADSET ||
                type == AudioDeviceInfo.TYPE_USB_HEADSET ||
                type == AudioDeviceInfo.TYPE_USB_DEVICE
    }

    private fun AudioDeviceInfo.isWirelessHeadset(): Boolean {
        val isBle = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            type == AudioDeviceInfo.TYPE_BLE_HEADSET ||
                    type == AudioDeviceInfo.TYPE_BLE_SPEAKER
        } else {
            false
        }
        return type == AudioDeviceInfo.TYPE_BLUETOOTH_A2DP ||
                type == AudioDeviceInfo.TYPE_BLUETOOTH_SCO ||
                isBle
    }
}
